internal/backend/runtime/omni/controllers/omni/machine_request_set_status.go

```diff
@@ -292,6 +315,37 @@ func (h *machineRequestSetStatusHandler) reconcileTearingDown(ctx context.Contex
 	return nil
 }
 
+func (h *machineRequestSetStatusHandler) handleInfraProviderDeletion(ctx context.Context, r controller.ReaderWriter, machine *infra.MachineRequest) error {
+	infraProviderID, ok := machine.Metadata().Labels().Get(omni.LabelInfraProviderID)
+	if !ok {
+		return nil
+	}
+
+	infraProvider, err := safe.ReaderGetByID[*infra.Provider](ctx, r, infraProviderID)
+	if err != nil && !state.IsNotFoundError(err) {
+		return err
+	}
+
+	// Remove all finalizers from all MachineRequests if the infra provider is being deleted or not found.
+	if infraProvider != nil && infraProvider.Metadata().Phase() == resource.PhaseRunning {
+		return nil
+	}
+
+	if err = safe.WriterModify[*infra.MachineRequest](ctx, r, machine, func(res *infra.MachineRequest) error {
```

Comment:
If we allow direct `MachineRequests`, we'll need to do the same in the `infra_provider_cleanup.go` controller properly.
For `MachineRequests` without the owner.